### PR TITLE
feat: new and initialize commands for migration tool

### PR DESCRIPTION
**Pull request by jzaralim**
### Description 
new: creates a new directory and empty properties file
initialize: creates migration stream and table

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

**Review by vcrfxia — COMMENTED** [review]
Thanks @jzaralim -- awesome stuff!! Comments inline ; nothing major.

File: ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/InitializeMigrationCommand.java
Comment (colinhicks) [reply ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/InitializeMigrationCommand.java:41]: Thanks for bringing this up. I think it's safe to assume customers running older Confluent Platform versions be interested in using this tool. For example, I think it's safe to assume someone upgrading to Confluent Platform 6.1.0 this month will try to use the migrations tool when it is released from the standalone distribution. We can and should be more specific about supported versions – let's continue this part of the discussion separately.

My assumption is that using `TIMESTAMP` would be more trouble than it's worth if we had to start backporting the implementation to, for example, 6.0.x and up. Is this correct? What do we lose in the long term by using `STRING` indefinitely instead?

File: ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/InitializeMigrationCommand.java
Comment (colinhicks) [reply ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/InitializeMigrationCommand.java:87]: +1

File: ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/InitializeMigrationCommand.java
Comment (agavra) [reply ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/InitializeMigrationCommand.java:84]: can we use log4j instead of stdout for logging? if users want to use stdout they can always configure the logger to do that.

File: ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/NewMigrationCommand.java
Comment (agavra) [reply ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/NewMigrationCommand.java:42]: +1 not only that, we should return correct error codes (using `System.exit`) if necessary so that it can be properly piped into scripts. I think getting error messages correct is a huge part of making this tool successful.

cc @colinhicks @spena  for their opinion on how this should work.

Comment (agavra) [reply ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/NewMigrationCommand.java:42]: I'd also recommend avoiding error stacks in the output that users see (logs can have the error messages). This turns people off from these tools and makes them look very rough around the edges.

Comment (colinhicks) [reply ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/NewMigrationCommand.java:42]: Yes, very important to return the correct error codes from the process. This is relevant for CI, too, where a non-zero code is expected to signal job step failure. We want to avoid forcing users to otherwise detect failure by parsing the output.

File: ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/InitializeMigrationCommand.java
Comment (vcrfxia) [reply ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/InitializeMigrationCommand.java:84]: Is this considered logging? This output should appear in the console. Otherwise `migrations initialize --dry-run` wouldn't print anything.

File: ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/InitializeMigrationCommand.java
Comment (vcrfxia) [reply ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/InitializeMigrationCommand.java:41]: Thanks for clarifying. Use of the Java client for pull queries requires ksqlDB server version of at least 0.10 (CP 6.0). Are we OK with the migrations tool being incompatible with earlier server versions?

Comment (vcrfxia) [reply ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/InitializeMigrationCommand.java:41]: Also, is there a reason we'd want to backport this to 6.0.x rather than having users use the tool released with a newer version (even if there server version is 6.0)? There was Java client functionality added post-6.0 (specifically, ksqlDB 0.12) for listing streams/tables that we might like to use in the migrations tool. (It's not a hard requirement so we can avoid it if backporting to 6.0.x is a requirement but I'd like to understand why that would be beneficial.)

Comment (jzaralim) [reply ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/InitializeMigrationCommand.java:41]: `CREATE OR REPLACE` was released in 6.1, so wouldn't this tool not even work for 6.0.x?

Comment (vcrfxia) [reply ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/InitializeMigrationCommand.java:41]: Why do we need `CREATE OR REPLACE`?

Comment (jzaralim) [reply ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/InitializeMigrationCommand.java:41]: I guess it would work, but I'd imagine updating schemas would be a major use case.

Comment (vcrfxia) [reply ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/InitializeMigrationCommand.java:41]: Are you saying that the migrations tool itself needs `CREATE OR REPLACE`, or that users will likely want `CREATE OR REPLACE` in the statements in their migrations files? As long as the migrations tool itself doesn't need `CREATE OR REPLACE`, then the tool is compatible with earlier server versions that do not support `CREATE OR REPLACE`, and can still be used to manage clusters of ksqlDB servers on those older versions. There have been a number of users in the community Slack requesting this feature while operating older ksqlDB server versions so I think that'd still be valuable. You're correct that their sql files wouldn't be able to contain `CREATE OR REPLACE`, but if they're starting from a version that doesn't support it then that should be the case anyway.

Comment (colinhicks) [reply ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/InitializeMigrationCommand.java:41]: > The best we can do if we want to use the Java client is keep the tool on master (and 6.1.x if desired) and announce compatibility with servers 6.0.x and later.

I think this is reasonable. Thanks for thinking through the compatibility implications.

File: ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/InitializeMigrationCommand.java
Comment (spena) [inline ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/InitializeMigrationCommand.java:73]: All commands will use this migrations configuration file. Can we avoid duplicating this call? Perhaps with something like:
```
final MigrationConfig properties = MigrationConfig.load();
```
^ That way, if the file does not exist , then `load` can return empty; or it can return some default values.

File: ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/MigrationConfig.java
Comment (spena) [inline ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/MigrationConfig.java:35]: Being `MigrationsConfig` an immutable object, and that only returns configs, I think we could rename the getter methods for simple names, like:
```
config.streamName();
config.streamTopic();
config.topicReplicas();
config.tableName();
config.tableTopic();
```

What do you think?

File: ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/InitializeMigrationCommand.java
Comment (spena) [reply ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/InitializeMigrationCommand.java:87]: +1

File: ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/InitializeMigrationCommand.java
Comment (spena) [reply ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/InitializeMigrationCommand.java:84]: Dry-run should display the same output as if no dry-run option was specified, with an extra message (at the beginning or end) that this was only a dry-run execution and no changes have been done in their ksql schema.

i.e. 
```
$ ksql-migrations initialize --dry-run
NOTICE: This execution is a dry-run. The ksql schema history stream and table will not be updated.

Creating stream: "migration_events"
Creating table: "schema_version"
Schema metadata initialized successfully (execution time 00:00.05s).
```

The statements that will be executed are too verbose imho. Are those statements going to help the user in some way? I think I agree with @agavra that we should log them as `info` logging.

File: ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/InitializeMigrationCommand.java
Comment (spena) [inline ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/InitializeMigrationCommand.java:112]: I recommend this output for the tool:
```
$ ksql-migrations initialize
Creating stream: "migration_events"
Successfully created migration stream.
Creating table: "schema_version"
Successfully created migration table.

Schema metadata initialized successfully (execution time 00:00.05s).
```
We should add the execution time at the end. This lets users know how much time their migration takes. 
We should print the action prior to executing it so in case the tool is stuck in a request, the user knows where. Otherwise, they will see an empty output and not knowing if the tool is making progress.

If it fails, then we could display:
```
$ ksql-migrations initialize
Creating stream: "migration_events"
Failed to created stream: blablablablabla
```

File: ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/InitializeMigrationCommand.java
Comment (spena) [reply ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/InitializeMigrationCommand.java:102]: +1

File: ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/NewMigrationCommand.java
Comment (spena) [inline ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/NewMigrationCommand.java:41]: Add more message about the progress of the tool.

i.e.
```
$ ksql-migrations new <project-dir>
Creating directory: <project-dir>
Creating directory: <project-dir>/migrations
Creating file: <project-dir>/ksql-migrations.properties

Migrations project directory created successfully (execution time 00:00.01s).
```

File: ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/NewMigrationCommand.java
Comment (spena) [inline ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/NewMigrationCommand.java:40]: The new config should have some default properties, and perhaps commented properties names in case the user wants to customize it.

File: ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/NewMigrationCommand.java
Comment (spena) [reply ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/NewMigrationCommand.java:42]: +1. I  also usually avoid putting everything into one try/catch when I want to know what failed. This way I can display something like:
```
$ ksql-migrations new /unknown-dir/migrations-project
Creating directory: <project-dir>
Failed creating directory: "Directory /unknown-dir does not exist"
```
Which then you can return know what error code to return, like `System.exit(2)`. This number will let user know what's wrong when writing scripts.

When working with user tools, it is always recommended to display a good error message and an error status. They will thank you a lot if they write their on bash scripts to automate this tool.

File: ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/NewMigrationCommand.java
Comment (spena) [inline ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/NewMigrationCommand.java:40]: We should check these do not exist before attempting to create them.

File: ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/InitializeMigrationCommand.java
Comment (vcrfxia) [reply ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/InitializeMigrationCommand.java:41]: Thanks, Colin. Sounds like we're all in agreement then.

> What do we lose in the long term by using STRING indefinitely instead?

We don't lose much, just that users might think it's odd that we're storing timestamps in the `STRING` type even though ksqlDB supports `TIMESTAMP`, and we're not creating visibility about the new type for users who don't know about it. This is minor compared to compatibility with older server versions so let's stick with the `STRING` type.

File: ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/InitializeMigrationCommand.java
Comment (vcrfxia) [reply ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/InitializeMigrationCommand.java:84]: Does dry-running initialize do anything (e.g., would dry-run fail if the metadata stream/table already exists) or does it always print that message regardless of the state of the cluster? If the latter, is it useful for `initialize` to support dry run?

I guess maybe you're saying the useful part is that it lets you know that it is going to try to create a stream and a table (and the names of the stream and table), and that the details of the stream and table are uninteresting to the user? I could be on board with that.

File: ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/NewMigrationCommand.java
Comment (vcrfxia) [reply ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/NewMigrationCommand.java:40]: +1 but I was going to suggest this as a follow-up. I think we should update the `new` command to take in the ksqlDB server URL, which we can then pre-populate into the config file. This is the only required config in the config file, so we definitely want it.

I don't feel strongly about whether we populate some of the other configs with defaults or not. Could be nice for some of the ones we expect users are more likely to want to change.

File: ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/InitializeMigrationCommand.java
Comment (spena) [reply ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/InitializeMigrationCommand.java:41]: Adding to this. Whether we use timestamp or string, I think we should print the string format when displaying time info in the command output. like `ksql-migrations info`.

File: ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/InitializeMigrationCommand.java
Comment (spena) [reply ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/InitializeMigrationCommand.java:84]: I'm not sure how useful is dry-run in the initialize, either. At least, one thing I find useful is if the command fails when creating the table (which is created after the stream). In that case, the stream wasn't created 'cause the dry-run option. Users would see their ksql cluster has some issues with the internal migrations stream/table without modifying anything on it.

But, is that a nice option to have? I don't think we need it in the initialize. @colinhicks what do you think? should we support dry-run only during a migration, or also on any command that modify the schema stream/table?

Comment (colinhicks) [reply ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/InitializeMigrationCommand.java:84]: Yeah `--dry-run` does not seem particularly useful here. I vote for dropping it from this command.

File: ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/NewMigrationCommand.java
Comment (jzaralim) [reply ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/NewMigrationCommand.java]: Done. For now, they all have the same error code, but after more parts of the tool are complete we can go through the errors and categorize them.

File: ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/MigrationConfig.java
Comment (jzaralim) [reply ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/MigrationConfig.java]: Extended `MigrationConfig` from `AbstractConfig` instead, so all these getter functions are removed.

File: ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/InitializeMigrationCommand.java
Comment (jzaralim) [reply ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/InitializeMigrationCommand.java]: Done, but for now `KsqlRestClient` is used in `MigrationConfig` to query the `/info` endpoint. This is temporary until the Java client supports this functionality.

File: ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/NewMigrationCommand.java
Comment (jzaralim) [reply ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/NewMigrationCommand.java]: Agreed, but I'll leave this as a follow-up

File: ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/MigrationConfig.java
Comment (vcrfxia) [inline ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/MigrationConfig.java:91]: This description is incomplete. Same for `KSQL_MIGRATIONS_TABLE_TOPIC_NAME` below.

File: ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/MigrationConfig.java
Comment (vcrfxia) [inline ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/MigrationConfig.java:133]: Don't worry about this for now since this code is going to be replaced with Java client code anyway, but after the replacement we should verify that this error message is actionable. For example, the current error message doesn't give any information regarding why the failure occurred (auth issue, transient network issue, invalid address, etc).

File: ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/InitializeMigrationCommand.java
Comment (vcrfxia) [inline ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/InitializeMigrationCommand.java:74]: We should refactor this time-tracking into BaseCommand since it's common to all commands. Feel free to postpone to a follow-up if you prefer.

File: ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/InitializeMigrationCommand.java
Comment (vcrfxia) [inline ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/InitializeMigrationCommand.java:104]: Creating the Java client from the MigrationsConfig should be factored out into a util file, since many more commands will need this. Feel free to defer to a follow-up PR.

File: ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/NewMigrationCommand.java
Comment (vcrfxia) [inline ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/NewMigrationCommand.java:83]: This log message should at least be a warning. It might make sense to even fail the command if the config file already exists, since it doesn't make sense for the user to run `migrations new` in this case, especially when we introduce a new param to specify the server URL as part of the command. (cc @colinhicks for a minor product question, in case he has input)

File: ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/NewMigrationCommand.java
Comment (vcrfxia) [inline ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/NewMigrationCommand.java:90]: This is missing `return false;` afterwards.

File: ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/MigrationConfig.java
Comment (vcrfxia) [inline ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/MigrationConfig.java:49]: What log4j config is being used with these loggers by default? Do we need to provide one with the tool?

File: ksqldb-tools/pom.xml
Comment (vcrfxia) [inline ksqldb-tools/pom.xml:71]: nit: spacing.

**Review by vcrfxia — APPROVED** [review]
Thanks @jzaralim -- looking great! A few suggestions and minor fixes inline. Feel free to address as follow-ups. LGTM assuming everything gets addressed.

File: ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/NewMigrationCommand.java
Comment (spena) [inline ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/NewMigrationCommand.java:76]: Shouldn't all these messages go to the command output? If an error occurs, the command will not show anything unless the user goes to the log.

File: ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/NewMigrationCommand.java
Comment (spena) [reply ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/NewMigrationCommand.java:83]: Shouldn't all these messages go to the command output? If the command fails, it just exits with `System.exit(1)`, and nothing shows the user what was wrong unless they go to the log.

File: ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/NewMigrationCommand.java
Comment (spena) [inline ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/NewMigrationCommand.java:49]: I am confused. Why is the message sent to the logger and not the command output?

File: ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/NewMigrationCommand.java
Comment (jzaralim) [reply ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/NewMigrationCommand.java]: changed to warn for now, might rethink this after adding the new param

File: ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/MigrationConfig.java
Comment (jzaralim) [reply ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/MigrationConfig.java:49]: Added a log4j config file that defaults to stdout

File: ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/NewMigrationCommand.java
Comment (jzaralim) [reply ksqldb-tools/src/main/java/io/confluent/ksql/tools/migrations/commands/NewMigrationCommand.java:49]: For now, the log4j config that we provide defaults to stdout. We might revisit this later.

**Pull request merged**
